pkg/infra/iac2/iac_templates.go

```diff
@@ -0,0 +1,121 @@
+package iac2
+
+import (
+	"context"
+	_ "embed"
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/query"
+	sitter "github.com/smacker/go-tree-sitter"
+	"github.com/smacker/go-tree-sitter/typescript/typescript"
+	"io"
+	"regexp"
+	"strings"
+	"text/template"
+)
+
+type (
+	ResourceCreationSignature struct {
+		inputTypes map[string]string
+		outputType string
+	}
+
+	ResourceCreationTemplate struct {
+		ResourceCreationSignature
+		imports            map[string]struct{}
+		expressionTemplate string
+	}
+)
+
+var (
+	tsLanguage = core.SourceLanguage{
+		ID:     core.LanguageId("ts"),
+		Sitter: typescript.GetLanguage(),
+	}
+
+	parameterizeArgsRegex = regexp.MustCompile(`args(\.\w+)`)
+
+	//go:embed find_args.scm
+	findArgsQuery string
+
+	//go:embed find_create_func.scm
+	findCreateFuncQuery string
+
+	//go:embed find_imports.scm
+	findImportsQuery string
+)
+
+func ParseResourceCreationTemplate(contents []byte) ResourceCreationTemplate {
```

Comment:
I'm curious why this approach instead of using Go templates like the previous iteration? I don't too much care for the "magic" of requiring `interface Args` and such over explicit template variables. Are there considerations I'm missing?

Reply:
ill let yuval explain more, but the reasoning was because we can then unit test and ensure that the ts files compile correctly

Reply:
Both approaches can be unit tested (but must be compiled first) so I don't see that as an advantage. Type-checking at build time is near use-less because they are pre-transformation so that has no bearing over whether the template is valid.

My main concern is that each transformation in this approach takes a lot of effort (multi-line queries, multi-line glue code in Go) compared to just `{{.Foo}}`.

Reply:
> Type-checking at build time is near useless

We have had _many_ bugs (granted, mostly in our runtime) due to typos and similar in non-go code that we embed. So I do think even just "aws.lambda.Function(...) requires certain args, do we have the right ones?" is useful to check.

For things like "do we have the right dependencies", were you envisioning something like a piece of go code that takes a template, compiles it with some dummy data (which may be non-trivial, for required args) to produce a valid `.ts` file, and then execs out to `npm install && npx tsc` to check that it's valid?

Another part of the testing that I forgot to mention is how the various resources fit together. For example, let's say you have a `Foo` resource whose template creates a `pulumi.Output<Foo>`. Then you have a `Bar` resource that takes in a `Foo` arg. It'd be nice if we could tell at build/test time that these things don't match up:

```go
type (
  Foo struct {}  // template produces a pulumi.Output<Foo>
  Bar struct {
    Foo *Foo // template requires a Foo, not pulumi.Output<Foo>
  }
)
```

If the templates only use `{{.Foo}}`, I think figuring that out is pretty hard. Whereas, if they're wrapped in this function, then it's pretty easy in a test to figure out:

* `Foo`'s template will generate a `"plumi.Output<Foo>"`
* `Bar.Foo` is a `Foo`, but its template requires a `"Foo"`
* `"pulumi.Output<Foo>" != "Foo"`, so test fails

Reply:
> We have had many bugs (granted, mostly in our runtime) due to typos and similar in non-go code that we embed.

No matter which approach, we can't make any guarantees that post-transformation will compile (or work) based on pre-transformation evaluations. Thus, we must make tests (unit or otherwise) for these in post-transformation. Given that, I don't see much incremental value in pre-transformation validation or testing.

> For things like "do we have the right dependencies", were you envisioning something like a piece of go code that takes a template, compiles it with some dummy data (which may be non-trivial, for required args) to produce a valid .ts file, and then execs out to npm install && npx tsc to check that it's valid?

Something like that yeah. Could even be non-trivial and run `jest` or `pulumi policy` on it (in addition to `tsc`). Could even be per-resource (baring dependencies). This kind of test is roughly what I had in mind for a post-transformation test.

> For example, let's say you have a Foo resource whose template creates a pulumi.Output<Foo>. Then you have a Bar resource that takes in a Foo arg.

`pulumi.Input` was designed for this purpose, so I'd propose we leverage that. Both `pulumi.Ouput<Foo>` and `Foo` satisfy `pulumi.Input<Foo>`. 

In the current approach, how would it resolve this? If one function returns `string` (say the createImage), and another uses `pulumi.Output<string>`, would it happen somewhere in the caller side to convert?

> If the templates only use {{.Foo}}, I think figuring that out is pretty hard. Whereas, if they're wrapped in this function, then it's pretty easy in a test to figure out.

Figuring that out from template side might be hard, but wouldn't the current approach have the same problems? You'd have to look at both templates to figure out that their types mismatch (except with the current approach, you'd also have to check the transformation code to make sure it didn't change anything either - unless we conventionally only transform a narrow portion of a template). I think both approaches are equally easy to test post-generation - both would fail a `tsc`.

Reply:
We talked offline. Summary: I didn't understand that these were "templates for templates" that are used to help write the templates in a way that is type checked and can carry the imports necessary. This approach is 💯 , just need some thought on template-level control flow (can be deferred until we need it).
Apologies on the churn ❤️ I should have read the whole thing before commenting 😄